### PR TITLE
Implement reference kernel for Add and Mul using CMSIS-NN

### PR DESCRIPTION
**Pull request by giorgio-arenarm**

### CODE REVIEW

**Comment by giorgio-arenarm** [comment]
@petewarden @gbaned gentle push for review

File: tensorflow/lite/micro/kernels/cmsis-nn/mul.cc [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+  double real_multiplier =
+      input1->params.scale * input2->params.scale / output->params.scale;
+  QuantizeMultiplier(real_multiplier, &data->output_multiplier,
+                     &data->output_shift);
+
+  return kTfLiteOk;
+}
+
+void* Init(TfLiteContext* context, const char* buffer, size_t length) {
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/cmsis-nn/mul.cc]: We can get rid of init altogether and instead register mul with nullptr for init below.

File: tensorflow/lite/micro/kernels/cmsis-nn/mul.cc [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  return nullptr;
+}
+
+void Free(TfLiteContext* context, void* buffer) {}
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/cmsis-nn/mul.cc]: same as above for init

File: tensorflow/lite/micro/kernels/cmsis-nn/mul.cc [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/cmsis-nn/mul.cc]: Same as init/free

File: tensorflow/lite/micro/kernels/cmsis-nn/mul.cc [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    case kTfLiteUInt8:
+    case kTfLiteInt8:
+      EvalQuantized(context, node, params, &data, input1, input2, output);
+      break;
+    case kTfLiteFloat32:
+      EvalFloat(context, node, params, &data, input1, input2, output);
+      break;
+    default:
+      context->ReportError(context, "Type %d not currently supported.",
+                           input1->type);
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/cmsis-nn/mul.cc]: For clearer type errors, please use:
context->ReportError(context, "Type %s (%d) not supported.", TfLiteTypeGetName(input->type), input->type);

**Review by njeffrie — CHANGES_REQUESTED** [review]

File: tensorflow/lite/micro/kernels/cmsis-nn/add.cc [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+                             OpData* data) {
+  data->requires_broadcast = !HaveSameShapes(input1, input2);
+
+  if (output->type == kTfLiteUInt8 || output->type == kTfLiteInt8) {
+    // 8bit -> 8bit general quantized path, with general rescalings
+    data->input1_offset = -input1->params.zero_point;
+    data->input2_offset = -input2->params.zero_point;
+    data->output_offset = output->params.zero_point;
+    data->left_shift = 20;
+    const double twice_max_input_scale =
```
[the remark this answers is not in the document]
Comment (giorgio-arenarm) [reply tensorflow/lite/micro/kernels/cmsis-nn/add.cc]: Ack, thank you for letting us know about that

File: tensorflow/lite/micro/kernels/cmsis-nn/mul.cc (hunk 1 above)
Comment (giorgio-arenarm) [reply tensorflow/lite/micro/kernels/cmsis-nn/mul.cc]: Done

File: tensorflow/lite/micro/kernels/cmsis-nn/mul.cc (hunk 3 above)
Comment (giorgio-arenarm) [reply tensorflow/lite/micro/kernels/cmsis-nn/mul.cc]: Done

File: tensorflow/lite/micro/kernels/cmsis-nn/mul.cc (hunk 4 above)
Comment (giorgio-arenarm) [reply tensorflow/lite/micro/kernels/cmsis-nn/mul.cc]: Done

File: tensorflow/lite/micro/kernels/cmsis-nn/mul.cc (hunk 2 above)
Comment (giorgio-arenarm) [reply tensorflow/lite/micro/kernels/cmsis-nn/mul.cc]: Done

**Review by njeffrie — APPROVED** [review]

**Pull request merged**